### DISCUSSION TITLE
How to write query to get Call Graph in JS?

### DISCUSSION THREAD

**Discussion by kal-purush** [github/codeql#9458] [Q&A]
is it possible to get a call graph for a codebase, preferably in dot format? I have found a few examples in other languages, so my guess is it should be possible but I can not figure it out.

**Comment by hmakholm** [comment]
This is not what you'd call an official feature, but I think something like this query ought to work:
```ql
import javascript

query predicate nodes(Function func, string key, string value) {
  key = "semmle.label" and value = func.getName()
}

from DataFlow::InvokeNode call, Function caller, Function callee
where caller = call.getEnclosingExpr().getEnclosingFunction()
  and callee = call.getACallee()
select caller, callee
```

You then need to run it with

     codeql query run -d DATABASE QUERY.ql -o TEMPFILE.bqrs

and after that run you can convert it to Dot with

    codeql bqrs interpret -t kind=graph --format dot TEMPFILE.bqrs -o OUTPUT-DIR

**Reply by kal-purush** [reply]
@hmakholm 
Thank you very much for the help. This works but I can not get the location(start line) of the caller or callee. If I add `call.getStartLine()` to select the query does not produce any result. Without that, it produces result.
